crates/trie/parallel/src/proof_task.rs

```diff
@@ -0,0 +1,320 @@
+    running_proof_tasks: Vec<RunningProofTask<Tx>>,
+    /// A receiver for new proof tasks.
+    proof_task_rx: Receiver<ProofTaskMessage>,
+}
+
+impl<Tx> ProofTaskManager<Tx> {
+    /// Creates a new [`ProofTaskManager`] with the given max concurrency, creating that number of
+    /// cursor factories.
+    ///
+    /// Returns an error if the consistent view provider fails to create a read-only transaction.
+    pub fn new<Factory>(
+        executor: Handle,
+        view: ConsistentDbView<Factory>,
+        task_ctx: ProofTaskCtx,
+        max_concurrency: usize,
+        proof_task_rx: Receiver<ProofTaskMessage>,
+    ) -> ProviderResult<Self>
+    where
+        Factory: DatabaseProviderFactory<Provider: BlockReader, DB: Database<TX = Tx>>
+            + StateCommitmentProvider
+            + 'static,
+    {
+        // initialize proof task txs
+        let mut proof_task_txs = Vec::with_capacity(max_concurrency);
+        for _ in 0..max_concurrency {
+            let provider_ro = view.provider_ro()?;
+            let tx = provider_ro.into_tx();
+            proof_task_txs.push(ProofTaskTx::new(tx, task_ctx.clone()));
+        }
+
+        Ok(Self {
+            pending_proofs: VecDeque::new(),
+            executor,
+            proof_task_txs,
+            running_proof_tasks: Vec::new(),
+            proof_task_rx,
+        })
+    }
+}
+
+impl<Tx> ProofTaskManager<Tx>
+where
+    Tx: DbTx + 'static,
+{
+    /// Spawns the proof task on the executor, with the input multiproof targets.
+    ///
+    /// If a task cannot be spawned immediately, this will be queued for completion later.
+    pub fn queue_proof_task(
+        &mut self,
+        input: StorageProofInput,
+        sender: Sender<StorageProofResult>,
+    ) {
+        self.pending_proofs.push_back((input, sender));
+    }
+
+    /// Spawns the next queued proof task on the executor with the given input, if there are any
+    /// transactions available.
+    pub fn try_spawn_next(&mut self) {
+        let Some(proof_task_tx) = self.proof_task_txs.pop() else { return };
+
+        let Some((pending_proof, sender)) = self.pending_proofs.pop_front() else {
+            // if there are no targets to do anything with put the tx back
+            self.proof_task_txs.push(proof_task_tx);
+            return
+        };
+
+        let (tx, rx) = channel();
+        self.executor.spawn_blocking(move || {
+            let output = proof_task_tx.storage_proof(pending_proof);
+            let _ = tx.send(output);
+        });
+
+        self.running_proof_tasks.push(RunningProofTask::new(rx, sender));
+    }
+
+    /// Loops, managing the proof tasks, and sending new tasks to the executor.
+    pub fn run(mut self) {
+        loop {
+            // TODO: condvar for
+            // * either running or proof tasks have stuff
+            // * otherwise yield thread
+            let message = match self.proof_task_rx.try_recv() {
+                Ok(message) => match message {
```

Comment:
hmm, this task now has do advance more than one channel, making this loop a bit complex with try_recv.

isn't the flow of messages just
<- incoming request for storage proof with a tx
if < concurrency: pick available db tx or spawn that task without one and create one on the task
handle the storageproof request on the task and return the db tx

then this only has to loop over a single receiver

or are we then running into issues with generics if we need to carry the db tx in the prooftaskmessage enum

Reply:
The design currently uses 2 channels, one for incoming proof requests, and one for returning txs from finished storage proof tasks. If we want only one channel in this loop, we would need another way (other than message passing) to keep track of txs in-use / not in-use

Reply:
> we would need another way (other than message passing) to keep track of txs in-use / not in-use

why? we could also return the tx itself as part of the message?

Reply:
Oops, I think I didn't understand the suggestion here, to reuse the input sender for sending the tx back. Yes we can do that